### PR TITLE
Recover IDE from runtime exceptions in renderer process

### LINKED ISSUE #1476: Provide user a possibility to recover from runtime exceptions
Right now, we handle most exceptions in a fairly user-friendly way(ones that may occur during transpilation process and in redux thunks).

But some lead to "crash to the gray screen", not giving a user any error messages or even a chance to save currently opened project(like #1468). Those exceptions are caused by a **combination** of current state and some bug in selectors or rendering code. In that case, we can't just try rendering the same thing again, because that would lead to the same exception.

But instead of rendering *nothing*, we can provide some recovery options
- export current changes to a `.xodball`
- undo last action and try rendering again
- simulate opening project in the current state(that could help if crash was caused by some complex UI manipulations)

### PR DESCRIPTION
**Pull request by brusherru**
It closes #1476 

---
It does not catch any exceptions. However, if some error crashes the GUI, the `<Catcher>` component will catch the exception, recover the state of the IDE to the previous stable state and show the report, which might be shared with the developers on GitHub or XOD forum.

The report includes:
- error message with stack trace,
- component stack,
- last three dispatched actions with the message did the last action broke the IDE or not,
- the stable state, where the tokens (user grant and project API token) are replaced with the string `SECRET`

To test this PR you can check it out and then broke something. For example:
- broke some selectors
- throw an exception in the `render()` method of some component
- reverse commit https://github.com/xodio/xod/commit/f96fa6d2732f8eff6df0bd869b93fbf4b348e43d and reproduce one of the old bugs from the closed issue: https://github.com/xodio/xod/issues/1468

---
This PR also could introduce some bugs in the IDE behavior after being recovered (because of reconstructing `<App>` and its children, so their state will be lost, and they should resubscribe on DOM and IPC events and so on), so I'll be glad if you find something and report it here.
Anyway, I'm sure it's better to get a recovered state and possibility to save the project, then just get a crashed IDE. :)

### CODE REVIEW

**Review by nkrkv — APPROVED** [review]
:+1: LGTM

File: packages/xod-client/src/core/containers/Catcher.jsx [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+      // Error
+      '# ERROR',
+      this.state.error.stack,
+      '# COMPONENT STACK',
+      this.state.errorInfo.componentStack,
+      '\n\n',
+      '# LAST ACTIONS (max 3 in log)',
+      isActionErrored
+        ? lastOneCausedError
+        : '# BUT NONE OF THEM IS NOT THE REASON',
```
Comment (evgenykochetkov) [inline packages/xod-client/src/core/containers/Catcher.jsx:92]: Maybe 
```suggestion
        : '# BUT NONE OF THEM IS THE REASON',
```
is more correct?

Or I didn't quite get the point of this message.

File: packages/xod-client/src/core/containers/Catcher.jsx [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        className={cls('IdeCrashReport', {
+          'is-shaded': this.state.recovering,
+        })}
+      >
+        {recovering}
+        <div className="Message">
+          <CloseButton onClick={this.onClose} />
+          <h1>Something went wrong</h1>
+          <p>
+            First of all, save the project after recovering process complete.
```
Comment (evgenykochetkov) [inline packages/xod-client/src/core/containers/Catcher.jsx:137]:
```suggestion
            First of all, save the project after recovering process completes.
```

or 

```suggestion
            First of all, save the project after recovering process is complete.
```

**Review by evgenykochetkov — APPROVED** [review]
The overall user experience is very cool 👍🏻
I have some doubts about a couple of messages, but I might be wrong.

**Review by evgenykochetkov — APPROVED** [review]

**Pull request merged**